pallets/stream-payment/src/lib.rs

```diff
@@ -0,0 +1,477 @@
+
+            // Noop
+            if new_rate_per_time_unit == stream.rate_per_time_unit {
+                return Ok(().into());
+            }
+
+            // Ensure rate change is fair.
+            if origin == stream.source && new_rate_per_time_unit < stream.rate_per_time_unit {
+                return Err(Error::<T>::SourceCantDecreaseRate.into());
+            }
+
+            if origin == stream.target && new_rate_per_time_unit > stream.rate_per_time_unit {
+                return Err(Error::<T>::TargetCantIncreaseRate.into());
+            }
+
+            // Perform pending payment before changing rate.
+            Self::perform_stream_payment(stream_id, &mut stream)?;
+
+            // Emit event.
+            Pallet::<T>::deposit_event(Event::<T>::StreamRateChanged {
+                stream_id,
+                old_rate: stream.rate_per_time_unit,
+                new_rate: new_rate_per_time_unit,
+            });
+
+            // Update rate
+            stream.rate_per_time_unit = new_rate_per_time_unit;
+            Streams::<T>::insert(stream_id, stream);
+
+            Ok(().into())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// Behavior:
+        /// A stream payment consist of a locked deposit, a rate per unit of time and the
+        /// last time the stream was updated. When updating the stream, **at most**
+        /// `elapsed_time * rate` is unlocked from the source account and transfered to the target
+        /// account. If this amount is greater than the left deposit, the stream is considered
+        /// drained **but not closed**. The source can come back later and refill the stream,
+        /// however there will be no retroactive payment for the time spent as drained.
+        /// If the stream payment is used to rent a service, the target should pause the service
+        /// while the stream is drained, and resume it once it is refilled.
+        fn perform_stream_payment(
+            stream_id: T::StreamId,
+            stream: &mut StreamOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let now =
+                T::TimeProvider::now(&stream.time_unit).ok_or(Error::<T>::CantFetchCurrentTime)?;
+
+            if now == stream.last_time_updated {
+                return Ok(().into());
+            }
+
+            // If deposit is zero the stream is fully drained and there is nothing to transfer.
+            if stream.deposit.is_zero() {
+                stream.last_time_updated = now;
+                return Ok(().into());
+            }
+
+            let delta = now
+                .checked_sub(&stream.last_time_updated)
+                .ok_or(Error::<T>::TimeOverflow)?;
+
+            // We compute the amount due to the target according to the rate, which may be
+            // lowered if the stream deposit is lower.
+            let mut payment = delta
+                .checked_mul(&stream.rate_per_time_unit)
+                .ok_or(Error::<T>::CurrencyOverflow)?;
```

Comment:
Here we could saturate to Balance::MAX, if someone sets a rate that will overflow I would expect that stream to be depleted immediately.

Reply:
Also, add some tests setting the rate to `0` and `u128::MAX`.